File: leetcode/106-construct-binary-tree-from-inorder-and-postorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
  var pos = {};
  for (var i = 0; i < inorder.length; ++i)
    pos[inorder[i]] = i;

  function build(p, q, len) {
    if (0 === len) return null;
    if (1 === len) return new TreeNode(inorder[p]);

    var val = postorder[q + len - 1];
    var k = pos[val];

    var d = k - p;
    var par = new TreeNode(val);
    par.left = build(p, q, d);
    par.right = build(p + d + 1, q + d, len - d - 1);

    return par;
  }

  return build(0, 0, inorder.length);
};

File: leetcode/106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]

        :rtype: TreeNode
        """
        pos = {v: i for i, v in enumerate(inorder)}

        def build(p, q, n):
            if 0 == n: return None
            if 1 == n: return TreeNode(inorder[p])

            val = postorder[q+n-1]
            k = pos[val]

            d = k - p
            par = TreeNode(val)
            par.left = build(p, q, d)
            par.right = build(p+d+1, q+d, n-d-1)

            return par

        return build(0, 0, len(inorder))
